### PR TITLE
Port real provision scenario for webui

### PR DESCRIPTION
**Pull request by oshtaier**
```
py.test /home/ashtayer/Desktop/robottelo/tests/foreman/ui_airgun/test_host.py -k 'test_positive_provision_end_to_end or test_positive_delete_libvirt'
==================================================================================== test session starts =====================================================================================
platform linux -- Python 3.6.5, pytest-4.0.1, py-1.5.3, pluggy-0.8.0
shared_function enabled - OFF - scope:  - storage: file
rootdir: /home/ashtayer/Desktop/robottelo, inifile:
plugins: services-1.2.1, mock-1.6.3
collecting ... 2019-05-20 16:34:21 - conftest - DEBUG - BZ deselect is disabled in settings

collected 31 items / 29 deselected                                                                                                                                                           

tests/foreman/ui_airgun/test_host.py ..                                                                                                                                                [100%]
=================================================================== 2 passed, 29 deselected, 3 warnings in 942.73 seconds ===================================================================
```
```
py.test /home/ashtayer/Desktop/robottelo/tests/foreman/ui_airgun/test_contentcredential.py::test_positive_search_scoped
==================================================================================== test session starts =====================================================================================
platform linux -- Python 3.6.5, pytest-4.0.1, py-1.5.3, pluggy-0.8.0
shared_function enabled - OFF - scope:  - storage: file
rootdir: /home/ashtayer/Desktop/robottelo, inifile:
plugins: services-1.2.1, mock-1.6.3
collecting ... 2019-05-20 16:57:12 - conftest - DEBUG - BZ deselect is disabled in settings

collected 1 item                                                                                                                                                                             

tests/foreman/ui_airgun/test_contentcredential.py .                                                                                                                                    [100%]
=========================================================================== 1 passed, 1 warnings in 122.60 seconds ==========================================================================
```
```
py.test /home/ashtayer/Desktop/robottelo/tests/foreman/ui_airgun/test_syncplan.py::test_positive_search_scoped
==================================================================================== test session starts =====================================================================================
platform linux -- Python 3.6.5, pytest-4.0.1, py-1.5.3, pluggy-0.8.0
shared_function enabled - OFF - scope:  - storage: file
rootdir: /home/ashtayer/Desktop/robottelo, inifile:
plugins: services-1.2.1, mock-1.6.3
collecting ... 2019-05-20 17:03:01 - conftest - DEBUG - BZ deselect is disabled in settings

collected 1 item                                                                                                                                                                             

tests/foreman/ui_airgun/test_syncplan.py .                                                                                                                                             [100%]

================================================================================= 1 passed in 91.56 seconds ===============================================================================
```

### CODE REVIEW

File: tests/foreman/ui_airgun/test_host.py [hunk 1]
```diff
@@ -1297,3 +1550,111 @@ def test_positive_bulk_delete_host(session, module_loc):
...
+            'interfaces.interface.network_type': 'Physical (Bridge)',
+            'interfaces.interface.network': settings.vlan_networking.bridge,
+        })
+        name = u'{0}.{1}'.format(hostname, module_libvirt_domain.name)
+        assert session.host.search(name)[0]['Name'] == name
+        try:
+            for _ in range(25):
+                values = session.host.get_details(name)
+                if values['properties']['properties_table']['Build'] == 'Pending installation':
+                    sleep(30)
```
Comment (vijay8451) [inline tests/foreman/ui_airgun/test_host.py]: Just a query ...why to use `sleep` , can we not use `wait_for` here ?

Comment (oshtaier) [reply tests/foreman/ui_airgun/test_host.py]: I believe we can try to use `wait_for`, so will look to change the code when gather all the comments, because I need full infrastructure to make even one re-test

File: tests/foreman/ui_airgun/test_host.py [hunk 2]
```diff
@@ -134,6 +145,248 @@ def module_host_template(module_org, module_loc):
...
+    else:
+        # Create Libvirt compute-resource
+        computeresource = entities.LibvirtComputeResource(
+            provider=FOREMAN_PROVIDERS['libvirt'],
+            url=resource_url,
+            set_console_password=False,
+            display_type=u'VNC',
+            location=[module_loc],
+            organization=[module_org],
+        ).create()
```
Comment (omkarkhatavkar) [inline tests/foreman/ui_airgun/test_host.py]: Do you think we have to validate that compute resource created is valid here?

File: tests/foreman/ui_airgun/test_host.py [hunk 3]
```diff
@@ -134,6 +145,248 @@ def module_host_template(module_org, module_loc):
...
+    repo.sync()
+    entity_mixins.TASK_TIMEOUT = old_task_timeout
+    return repo
+
+
+@pytest.fixture(scope='module')
+def module_libvirt_content_view(module_org, module_libvirt_repository, module_libvirt_lce):
+    # Create, Publish and promote CV
+    old_task_timeout = entity_mixins.TASK_TIMEOUT
+    entity_mixins.TASK_TIMEOUT = 3600
```
Comment (omkarkhatavkar) [inline tests/foreman/ui_airgun/test_host.py]:
```
old_task_timeout = entity_mixins.TASK_TIMEOUT
entity_mixins.TASK_TIMEOUT = 3600
```
set and uset timeout used multiple times, can we put that into a common function?

File: tests/foreman/ui_airgun/test_host.py (hunk 2 above)
Comment (oshtaier) [reply tests/foreman/ui_airgun/test_host.py]: Nope, as:
1) All these preconditions are very strict. Even if one element goes wrong, the whole scenario will fail. For example:
    a) if compute resource is not valid, host 'Virtual Machine' tab will produce an exception and corresponding field will not be accessible. As result, host will not be created and we catch the problem
    b) if your subnet is not valid, your DHCP server will not provide ip from the pool and you get an exception for your network interface.
    c) if your domain is not valid, you will also get an exception as subnet will not be accessible
2) Such validations were not present in the initial scenario
3) It is API functionality which is always out of scope for WebUI validation


Basically, it makes sense to improve something from WebUI side, like return exact exception rather than fact that we cannot create the host, but it is low priority as we should see the error on a screenshot

File: tests/foreman/ui_airgun/test_host.py [hunk 4]
```diff
@@ -134,6 +145,248 @@ def module_host_template(module_org, module_loc):
...
+            mask=settings.vlan_networking.netmask,
+            location=[module_loc],
+            organization=[module_org],
+            domain=[module_libvirt_domain],
+            ipam='DHCP',
+            dns=module_proxy,
+            dhcp=module_proxy,
+            tftp=module_proxy,
+            discovery=module_proxy,
+        ).create()
```
[the remark this answers is not in the document]
Comment (oshtaier) [reply tests/foreman/ui_airgun/test_host.py]: explained above

File: tests/foreman/ui_airgun/test_host.py [hunk 5]
```diff
@@ -134,6 +145,248 @@ def module_host_template(module_org, module_loc):
...
+        domain.location.append(module_loc)
+        domain.organization.append(module_org)
+        domain.dns = module_proxy
+        domain = domain.update(['dns', 'location', 'organization'])
+    else:
+        domain = entities.Domain(
+            dns=module_proxy,
+            location=[module_loc],
+            organization=[module_org],
+        ).create()
```
[the remark this answers is not in the document]
Comment (oshtaier) [reply tests/foreman/ui_airgun/test_host.py]: explained above

File: tests/foreman/ui_airgun/test_host.py [hunk 6]
```diff
@@ -134,6 +145,248 @@ def module_host_template(module_org, module_loc):
...
+    # Create new compute-resource with 'libvirt' provider.
+    resource_url = u'qemu+ssh://root@{0}/system'.format(
+        settings.compute_resources.libvirt_hostname
+    )
+    comp_res = [
+        res for res in entities.LibvirtComputeResource().search()
+        if res.provider == FOREMAN_PROVIDERS['libvirt'] and res.url == resource_url
+    ]
+    if len(comp_res) > 0:
+        computeresource = entities.LibvirtComputeResource(id=comp_res[0].id).read()
```
[the remark this answers is not in the document]
Comment (oshtaier) [reply tests/foreman/ui_airgun/test_host.py]: sure

File: tests/foreman/ui_airgun/test_host.py (hunk 3 above)
Comment (oshtaier) [reply tests/foreman/ui_airgun/test_host.py]: Did you see that we do?:
```
timeout = new
doing some functionality
timeout =old
```
not sure how to make it generic. Do you propose to create some common function that will receive functions as a parameter. But we can have multiple statements, so lambda or similar ideology doesn't fit our needs.

File: tests/foreman/ui_airgun/test_host.py (hunk 2 above)
Comment (omkarkhatavkar) [reply tests/foreman/ui_airgun/test_host.py]: As all of you module fixtures are part of test settings any failure is not recorded immediately and wait for the next fixture to show if that is failed. While analysing test, it will be each fixture depend on others to show the result. Probably that will make the user search in the wrong side of failure as the test data is invalid for my next fixture.

I see you point 2 & 3 valid enough. But consider a case where All Test setting got passed as there is no validation or assert, UI test will start and ultimately get failed ( This is time consuming based on UI test ). That was my whole point.

File: tests/foreman/ui_airgun/test_host.py (hunk 3 above)
Comment (omkarkhatavkar) [reply tests/foreman/ui_airgun/test_host.py]: That's OK then. I'm fine with it.

File: tests/foreman/ui_airgun/test_contentcredential.py [hunk 7]
```diff
@@ -92,6 +92,33 @@ def test_positive_end_to_end(session, module_org, gpg_content):
...
+
+    :CaseImportance: High
+    """
+    name = gen_string('alpha')
+    org = entities.Organization().create()
+    with session:
+        session.organization.select(org.name)
+        session.contentcredential.create({
+            'name': name,
+            'content_type': 'GPG Key',
```
Comment (ldjebran) [inline tests/foreman/ui_airgun/test_contentcredential.py]: Think we need to create a constant for content_credentials types
```python
CONTENT_CREDENTIALS_TYPES = {
    'gpg': ''GPG Key'',
    'ssl': 'SSL Certificate'
}
```
and use it here

```python
 'content_type': CONTENT_CREDENTIALS_TYPES['gpg'], 

```

File: tests/foreman/ui_airgun/test_host.py (hunk 2 above)
Comment (oshtaier) [reply tests/foreman/ui_airgun/test_host.py]: Once more, how do you propose to validate? `create()` already doing `read()`, so if something goes wrong with create, you will get an exception anyway. But to understand that entity is really valid, assert doesn't suffice our needs and we actually can validate that fact on host create procedure only

File: tests/foreman/ui_airgun/test_host.py [hunk 8]
```diff
@@ -134,6 +145,248 @@ def module_host_template(module_org, module_loc):
...
+def default_architecture():
+    # Get the architecture ID
+    return entities.Architecture().search(
+        query={u'search': u'name="{0}"'.format(DEFAULT_ARCHITECTURE)})[0]
+
+
+@pytest.fixture(scope='module')
+def module_libvirt_environment(module_org, module_loc):
+    # Search for puppet environment and associate location
+    environment = entities.Environment(organization=[module_org]).search()[0]
```
Comment (ldjebran) [inline tests/foreman/ui_airgun/test_host.py]: I'am not sure that this will return the environment related to organization (it can return all the environments) 
to be more sure use
```python
entities.Environment().search(query=dict(search="organization_id={}".format(module_org.id)))[0]
```

File: tests/foreman/ui_airgun/test_host.py (hunk 2 above)
Comment (omkarkhatavkar) [reply tests/foreman/ui_airgun/test_host.py]: OK Valid enough.

File: tests/foreman/ui_airgun/test_host.py (hunk 8 above)
Comment (oshtaier) [reply tests/foreman/ui_airgun/test_host.py]: According to:
https://nailgun.readthedocs.io/en/latest/examples.html#search

>If any attributes have been set, they are used. This finds all lifecycle environments that have a name of “foo” and that belong to organization 1:
`lc_envs = LifecycleEnvironment(name='foo', organization=1).search()`

File: tests/foreman/ui_airgun/test_host.py [hunk 9]
```diff
@@ -134,6 +145,248 @@ def module_host_template(module_org, module_loc):
...
+def module_libvirt_product(module_org):
+    return entities.Product(organization=module_org).create()
+
+
+@pytest.fixture(scope='module')
+def module_libvirt_repository(os_path, module_libvirt_product):
+    old_task_timeout = entity_mixins.TASK_TIMEOUT
+    repo = entities.Repository(
+        product=module_libvirt_product, url=os_path).create()
+    entity_mixins.TASK_TIMEOUT = 3600
```
Comment (ldjebran) [inline tests/foreman/ui_airgun/test_host.py]: use call_entity_method_with_timeout from api.utils
```python
call_entity_method_with_timeout(repo.sync, timeout=3600)
```

File: tests/foreman/ui_airgun/test_host.py (hunk 3 above)
Comment (ldjebran) [reply tests/foreman/ui_airgun/test_host.py]: we cannot be sure that the TASK_TIEOUT is restored  in case of failure we can use
call_entity_method_with_timeout for each function as commented above or 
```python
old_task_timeout = entity_mixins.TASK_TIMEOUT
entity_mixins.TASK_TIMEOUT = 3600
try:
    ....

finally: 
     entity_mixins.TASK_TIMEOUT = old_task_timeout

```

**Comment by san7ket** [comment]
Can't we use https://github.com/SatelliteQE/robottelo/blob/master/robottelo/api/utils.py#L291. It returns all the entities required for provisioning

**Comment by oshtaier** [comment]
Yeah, we created exactly that. All fixtures here are independent from each other, so user can easily substitute only needed entities, so all types of provisioning can be used here

**Comment by oshtaier** [comment]
```
py.test /home/ashtayer/Desktop/robottelo/tests/foreman/ui_airgun/test_host.py -k 'test_positive_provision_end_to_end or test_positive_delete_libvirt'
==================================================================================== test session starts =====================================================================================
platform linux -- Python 3.6.5, pytest-4.0.2, py-1.5.3, pluggy-0.11.0
shared_function enabled - OFF - scope:  - storage: file
rootdir: /home/ashtayer/Desktop/robottelo, inifile:
plugins: services-1.3.1, mock-1.10.0
collecting ... 2019-05-28 15:24:33 - conftest - DEBUG - BZ deselect is disabled in settings

collected 31 items / 29 deselected                                                                                                                                                           

tests/foreman/ui_airgun/test_host.py ..                                                                                                                                                [100%]

======================================================================================2 passed, 29 deselected, 3 warnings in 1128.43 seconds ===================================================================
```

**Comment by oshtaier** [comment]
```
py.test /home/ashtayer/Desktop/robottelo/tests/foreman/ui_airgun/test_computeresource_libvirt.py 
==================================================================================== test session starts =====================================================================================
platform linux -- Python 3.6.5, pytest-4.0.2, py-1.5.3, pluggy-0.11.0
shared_function enabled - OFF - scope:  - storage: file
rootdir: /home/ashtayer/Desktop/robottelo, inifile:
plugins: services-1.3.1, mock-1.10.0
collecting ... 2019-05-27 20:14:06 - conftest - DEBUG - BZ deselect is disabled in settings

collected 1 item                                                                                                                                                                             

tests/foreman/ui_airgun/test_computeresource_libvirt.py .                                                                                                                              [100%]
=========================================================================== 1 passed, 7 warnings in 214.31 seconds ========================================================================
```

**Comment by oshtaier** [comment]
```
py.test /home/ashtayer/Desktop/robottelo/tests/foreman/ui_airgun/test_host.py -k 'test_positive_provision_end_to_end or test_positive_delete_libvirt'
==================================================================================== test session starts =====================================================================================
platform linux -- Python 3.6.5, pytest-4.0.2, py-1.5.3, pluggy-0.11.0
shared_function enabled - OFF - scope:  - storage: file
rootdir: /home/ashtayer/Desktop/robottelo, inifile:
plugins: services-1.3.1, mock-1.10.0
collecting ... 2019-05-28 17:27:56 - conftest - DEBUG - BZ deselect is disabled in settings

collected 31 items / 29 deselected                                                                                                                                                           

tests/foreman/ui_airgun/test_host.py ..                                                                                                                                                [100%]

====================================================================================== 
=================================================================== 2 passed, 29 deselected, 3 warnings in 1032.89 seconds ==================================================================
```
